protocols/gossipsub/src/peer_score.rs

```diff
@@ -225,16 +223,31 @@ impl PeerScore {
 
     /// Returns the score for a peer
     pub(crate) fn score(&self, peer_id: &PeerId) -> f64 {
-        self.metric_score(peer_id, None)
+        let (score, _, _) = self.calculate_score(peer_id);
+        score
     }
 
     /// Returns the score for a peer, logging metrics. This is called from the heartbeat and
     /// increments the metric counts for penalties.
-    pub(crate) fn metric_score(&self, peer_id: &PeerId, mut metrics: Option<&mut Metrics>) -> f64 {
+    #[cfg(feature = "metrics")]
+    pub(crate) fn metric_score(&self, peer_id: &PeerId, metrics: Option<&mut Metrics>) -> f64 {
+        let (score, num_message_deficit, num_ip_colocation) = self.calculate_score(peer_id);
+        if let Some(m) = metrics {
+            m.register_score_penalty(Penalty::MessageDeficit, num_message_deficit);
+            m.register_score_penalty(Penalty::IPColocation, num_ip_colocation);
+        }
+        score
+    }
+
+    fn calculate_score(&self, peer_id: &PeerId) -> (f64, u64, u64) {
         let Some(peer_stats) = self.peer_stats.get(peer_id) else {
-            return 0.0;
+            return (0.0, 0, 0);
         };
         let mut score = 0.0;
+        #[cfg(feature = "metrics")]
+        let mut num_message_deficit_penalties = 0;
+        #[cfg(feature = "metrics")]
+        let mut num_ip_colocation_penalties = 0;
```

Comment:
Don't think we need to feature gate them. 
Without metrics enabled they'll just stay `0`, and we can return them below.

Reply:
There will be warning about unused mutability, not very easy to silence that so I just gate them.

Reply:
Hmm, we could add something like `#[cfg_attr(not(feature= "metrics"), allow(unused_mut))]` to the `calculate_score` function, or prefix the var names with an `_` to silence the warning.